File: Swift/Tests/ParserTests/CalculatorTests.swift
import Testing
import Foundation

@testable import Parser

@Test func testFuncs(){
	let calc = Calculator()

	#expect(calc.calculate("sin(2.0)")! == (sin(2.0), ""))
	#expect(calc.calculate("cos(2.0)")! == (cos(2.0), ""))
	#expect(calc.calculate("asin(0.5)")! == (asin(0.5), ""))
	#expect(calc.calculate("acos(0.5)")! == (acos(0.5), ""))
	#expect(calc.calculate("sinh(2.0)")! == (sinh(2.0), ""))
	#expect(calc.calculate("cosh(2.0)")! == (cosh(2.0), ""))
	#expect(calc.calculate("asinh(2.0)")! == (asinh(2.0), ""))
	#expect(calc.calculate("acosh(2.0)")! == (acosh(2.0), ""))
	#expect(calc.calculate("tan(2.0)")! == (tan(2.0), ""))
	#expect(calc.calculate("log(2.0)")! == (log(2.0), ""))
	#expect(calc.calculate("log10(2.0)")! == (log10(2.0), ""))
	#expect(calc.calculate("exp(2.0)")! == (exp(2.0), ""))
	#expect(calc.calculate("sqrt(2.0)")! == (sqrt(2.0), ""))
	#expect(calc.calculate("sqr(2.0)")! == (4.0, ""))
}

@Test func testConsts(){
	let calc = Calculator()
	let pi = 3.14159265358979323846

	#expect(calc.calculate("E")! == (2.7182818284590452, ""))
	#expect(calc.calculate("LOG2E")! == (1 / log(2.0), ""))
	#expect(calc.calculate("LOG10E")! == (0.4342944819032518, ""))
	//#expect(calc.calculate("LOG10E")! == (1 / log(10.0), ""))
	#expect(calc.calculate("LN2")! == (log(2.0), ""))
	#expect(calc.calculate("LN10")! == (log(10.0), ""))
	#expect(calc.calculate("PI")! == (pi, ""))
	#expect(calc.calculate("PI_2")! == (pi / 2, ""))
	#expect(calc.calculate("PI_4")! == (pi / 4, ""))
	#expect(calc.calculate("1_PI")! == (1 / pi, ""))
	#expect(calc.calculate("2_PI")! == (2 / pi, ""))
	#expect(calc.calculate("2_SQRTPI")! == (2 / sqrt(pi), ""))
	#expect(calc.calculate("SQRT2")! == (sqrt(2), ""))
	#expect(calc.calculate("SQRT1_2")! == (sqrt(0.5), ""))
}

@Test func testCalculator(){
	let calc = Calculator()

	#expect(calc.calculate("72 - 7 - (1 - 2) * 3")! == (68.0, ""))
	#expect(calc.calculate("-72 - 7 - (1 - 2) * 3")! == (-76.0, ""))
	#expect(calc.calculate(" 7.21e-1 - 7.3 - (1.5 - 2.2) * (-3.3)")! == (-8.889, ""))
	#expect(calc.calculate("3^(1+1)^3").map({ (round($0.0), $0.1) })! == (6561.0, ""))
	#expect(calc.calculate("sin(1+1)")! == (sin(2.0), ""))
	#expect(calc.calculate("sin ( 2_SQRTPI * sqr ( 2 ) - 1 )")! == (-0.3634085731426532, ""))
	#expect(calc.calculate("sqr(2 + 3)")! == (25.0, ""))
	#expect(calc.calculate("sin(-PI/4)")! == (-0.7071067811865475, ""))
	#expect(calc.calculate(" E ^ PI")! == (23.140692632779267, ""))
	#expect(calc.calculate(" PI ^ E")! == (22.45915771836104, ""))
}

File: java/src/test/java/kmm/parser/CalculatorTests.java
package kmm.parser;

import kmm.utils.Pair;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class CalculatorTests {
    @Test
    public void test_funcs() {
        Parser<Double> pexpr = Calculator.expr();
        assertEquals(Optional.of(Pair.of(Math.sin(2.0), "")), pexpr.parse("sin(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.cos(2.0), "")), pexpr.parse("cos(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.asin(2.0), "")), pexpr.parse("asin(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.acos(2.0), "")), pexpr.parse("acos(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.sinh(2.0), "")), pexpr.parse("sinh(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.cosh(2.0), "")), pexpr.parse("cosh(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.tan(2.0), "")), pexpr.parse("tan(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.log(2.0), "")), pexpr.parse("log(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.log10(2.0), "")), pexpr.parse("log10(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.exp(2.0), "")), pexpr.parse("exp(2.0)"));
        assertEquals(Optional.of(Pair.of(Math.sqrt(2.0), "")), pexpr.parse("sqrt(2.0)"));
        assertEquals(Optional.of(Pair.of(4.0, "")), pexpr.parse("sqr(2.0)"));
    }

    @Test
    public void test_consts() {
        Parser<Double> pexpr = Calculator.expr();
        assertEquals(Optional.of(Pair.of(Math.PI, "")), pexpr.parse("PI"));
        assertEquals(Optional.of(Pair.of(Math.E, "")), pexpr.parse("E"));
        assertEquals(Optional.of(Pair.of(1 / Math.log(2.0), "")), pexpr.parse("LOG2E"));
        assertEquals(Optional.of(Pair.of(0.4342944819032518, "")), pexpr.parse("LOG10E"));
        assertEquals(Optional.of(Pair.of(Math.log(2.0), "")), pexpr.parse("LN2"));
        assertEquals(Optional.of(Pair.of(Math.log(10.0), "")), pexpr.parse("LN10"));
        assertEquals(Optional.of(Pair.of(Math.PI / 2, "")), pexpr.parse("PI_2"));
        assertEquals(Optional.of(Pair.of(Math.PI / 4, "")), pexpr.parse("PI_4"));
        assertEquals(Optional.of(Pair.of(1 / Math.PI, "")), pexpr.parse("1_PI"));
        assertEquals(Optional.of(Pair.of(2 / Math.PI, "")), pexpr.parse("2_PI"));
        assertEquals(Optional.of(Pair.of(2 / Math.sqrt(Math.PI), "")), pexpr.parse("2_SQRTPI"));
        assertEquals(Optional.of(Pair.of(Math.sqrt(2), "")), pexpr.parse("SQRT2"));
        assertEquals(Optional.of(Pair.of(Math.sqrt(0.5), "")), pexpr.parse("SQRT1_2"));
    }

    @Test
    public void test_calculator() {
        Parser<Double> pexpr = Calculator.expr();
        assertEquals(Optional.of(Pair.of(0.4342944819032518, "")), pexpr.parse("LOG10E"));
        assertEquals(Optional.of(Pair.of(68.0, "")), pexpr.parse("72 - 7 - (1 - 2) * 3"));
        assertEquals(Optional.of(Pair.of(-8.889, "")), pexpr.parse(" 7.21e-1 - 7.3 - (1.5 - 2.2) * (-3.3)"));
        assertEquals(Optional.of(Pair.of(6561.0, "")), pexpr.parse("3^(1+1)^3"));
        assertEquals(Optional.of(Pair.of(Math.sin(2), "")), pexpr.parse("sin(1+1)"));
        assertEquals(Optional.of(Pair.of(-0.3634085731426532, "")), pexpr.parse("sin ( 2_SQRTPI * sqr ( 2 ) - 1 )"));
        assertEquals(Optional.of(Pair.of(25.0, "")), pexpr.parse("sqr(2 + 3)"));
        assertEquals(Optional.of(Pair.of(Math.pow(Math.E, Math.PI), "")), pexpr.parse("E ^ PI"));
        assertEquals(Optional.of(Pair.of(Math.pow(Math.PI, Math.E), "")), pexpr.parse("PI ^ E"));
    }
}
